### PR TITLE
<regex>: Fixes regex character class case insensitive search problem

### LINKED ISSUE #993: <regex>: c++ regex character class case insensitive search problem
**Describe the bug**
`regex::icase` is not handled correctly for some input.

**Command-line test case**
```
d:\Temp2>type repro.cpp
#include <iostream>
#include <regex>

using namespace std;

int main()
{
        wstring target(L" Copyright");
        wsmatch match;
        wregex regexp(L"[a-z][a-z]", regex::icase);
        if (regex_search(target.cbegin(), target.cend(), match, regexp)) {
                wcout << L"Matched: \"" << match.str() << L"\"" << endl;
        }
        return (0);
}
d:\Temp2>cl /EHsc /W4 /WX .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
Microsoft (R) Incremental Linker Version 14.27.29009.1
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe
repro.obj

d:\Temp2>.\repro.exe
Matched: "op"
```

**Expected behavior**
```
Matched: "Co"
```
**Command-line test case 2**
```
d:\Temp2>type repro.cpp
#include <regex>
#include <iostream>

int main()
{
        const wchar_t* test_raw_string = L"blahZblah";
        const wchar_t* test_patterns[] = { L"[Z]", L"[z]" };

        for (const auto* test_pattern : test_patterns)
        {
                const std::wstring test_string = test_raw_string;
                std::wregex case_regex(test_pattern, std::regex_constants::ECMAScript);
                std::wcmatch match1;

                std::wcout << test_pattern << L" search " << test_string << L" case sensitive = " << std::regex_search(test_string, case_regex) << L'\n';
                std::wcout << test_pattern << L" search " << test_string << L" case sensitive with match = " << std::regex_search(test_raw_string, match1, case_regex) << L'\n';

                std::wregex icase_regex(test_pattern, std::regex_constants::icase | std::regex_constants::ECMAScript);
                std::wcmatch match2;

                std::wcout << test_pattern << L" search " << test_string << L" case insensitive = " << std::regex_search(test_string, icase_regex) << L'\n';
                std::wcout << test_pattern << L" search " << test_string << L" case insensitive with match = " << std::regex_search(test_raw_string, match2, icase_regex) << L'\n';

                std::wcout << L'\n';
        }
    return 0;
}


d:\Temp2>cl /EHsc /W4 /WX .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
Microsoft (R) Incremental Linker Version 14.27.29009.1
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe
repro.obj

d:\Temp2>.\repro.exe
[Z] search blahZblah case sensitive = 1
[Z] search blahZblah case sensitive with match = 1
[Z] search blahZblah case insensitive = 0
[Z] search blahZblah case insensitive with match = 0

[z] search blahZblah case sensitive = 0
[z] search blahZblah case sensitive with match = 0
[z] search blahZblah case insensitive = 0
[z] search blahZblah case insensitive with match = 0
```
**Expected behavior 2**
```
[Z] search blahZblah case sensitive = 1
[Z] search blahZblah case sensitive with match = 1
[Z] search blahZblah case insensitive = 1
[Z] search blahZblah case insensitive with match = 1

[z] search blahZblah case sensitive = 0
[z] search blahZblah case sensitive with match = 0
[z] search blahZblah case insensitive = 1
[z] search blahZblah case insensitive with match = 1
```

**STL version**
```
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1
```

**Additional context**
This item is also tracked on Developer Community as DevCom-230267 and DevCom-246258 and by Microsoft-internal VSO-287844 / [AB#287844](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/287844).

See also #405

### PR DESCRIPTION
**Pull request by Arzaghi**
Fixes #993

### CODE REVIEW

File: tests/std/include/test_regex_support.hpp [hunk 1]
```diff
@@ -240,3 +240,64 @@ class test_regex {
...
+                } else {
+                    puts("but it failed to match");
+                }
+
+                fixture->fail_regex();
+            }
+        } catch (const std::regex_error& e) {
+            wprintf(LR"(Failed to regex_search("%s", regex("%s", 0x%X), 0x%X): regex_error: )", subject.c_str(),
+                pattern.c_str(), static_cast<unsigned int>(syntax), static_cast<unsigned int>(match_flags));
+            printf("%s\n", e.what());
```
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:280]: The narrow `test_regex` above printed `e.what()` in quotes. I believe it's preferable for `test_wregex` to emit the same style of output, although they have different ways of forming that output (I like your technique of simply using a narrow `printf` call for the narrow `e.what()` :smiley_cat:):
```suggestion
            printf("\"%s\"\n", e.what());
```

File: tests/std/include/test_regex_support.hpp [hunk 2]
```diff
@@ -240,3 +240,64 @@ class test_regex {
...
+                wprintf(LR"(Expected regex_search("%s", regex("%s", 0x%X), 0x%X) to not match, but it found "%s")"
+                        "\n",
+                    subject.c_str(), pattern.c_str(), static_cast<unsigned int>(syntax),
+                    static_cast<unsigned int>(match_flags), mr.str().c_str());
+                fixture->fail_regex();
+            }
+        } catch (const std::regex_error& e) {
+            wprintf(LR"(Failed to regex_search("%s", regex("%s", 0x%X), 0x%X): regex_error: )", subject.c_str(),
+                pattern.c_str(), static_cast<unsigned int>(syntax), static_cast<unsigned int>(match_flags));
+            printf("%s\n", e.what());
```
Comment (StephanTLavavej) [inline tests/std/include/test_regex_support.hpp:299]:
```suggestion
            printf("\"%s\"\n", e.what());
```

File: tests/std/tests/VSO_0000000_regex_use/test.cpp [hunk 3]
```diff
@@ -546,6 +546,34 @@ void test_VSO_226914_word_boundaries() {
...
+
+    {
+        const wstring subject = L"blahZblah";
+        const test_wregex Z_case_regex(&g_regexTester, LR"([Z])", ECMAScript);
+        const test_wregex Z_icase_regex(&g_regexTester, LR"([Z])", ECMAScript | icase);
+        const test_wregex z_case_regex(&g_regexTester, LR"([z])", ECMAScript);
+        const test_wregex z_icase_regex(&g_regexTester, LR"([z])", ECMAScript | icase);
+
+        Z_case_regex.should_search_match(subject, L"Z");
+        Z_icase_regex.should_search_match(subject, L"Z");
```
Comment (StephanTLavavej) [inline tests/std/tests/VSO_0000000_regex_use/test.cpp:567]: I'd like to see one more thing for comprehensive test coverage - verifying that `[Z]` matches `z` with `icase`.

**Review by StephanTLavavej — COMMENTED** [review]
Looks good, thanks for fixing this bug! I'll push one more test case and trivial stylistic changes to make `test_regex` and `test_wregex` even more similar.

**Review by StephanTLavavej — APPROVED** [review]

**Review by barcharcraz — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@Arzaghi @barcharcraz I've pushed a merge with `master` which applied clang-format 11 to this code. The added indentation caused annoying wrapping of the newly added line here, so I extracted the `_Uelem` typedef to make it more readable - no behavioral changes. (In fact I didn't bother to build/test, so let's see if I can get away with it...)

**Comment by StephanTLavavej** [comment]
THANKS For fixing ThIs cAsE INSENSITIVE Bug! :joy_cat:

**Pull request merged**